services/worker/src/worker/job_runners/config/parquet_and_info.py

```diff
@@ -660,8 +650,164 @@ def fill_builder_info(builder: DatasetBuilder, hf_token: Optional[str]) -> None:
+
+    def __enter__(self) -> "limit_parquet_writes":
+        limiter = self
+
+        class _TrackedParquetWriter(pq.ParquetWriter):  # type: ignore
+            """Count on-the-fly how many bytes are written"""
+
+            def track_write_table(self, pa_table: pa.Table) -> None:
+                limiter.total_bytes += pa_table.nbytes
+
+            def write_table(self, pa_table: pa.Table, row_group_size: Optional[int] = None) -> None:
+                self.track_write_table(pa_table)
+                super().write_table(pa_table, row_group_size=row_group_size)
+
+        def limited_generator(
+            generator: Callable[..., Generator[T, None, None]]
+        ) -> Callable[..., Generator[T, None, None]]:
+            """Stop the underlying generator once we reach the maximum dataset size"""
+
+            @functools.wraps(generator)
+            def wrapped(*args: Any, **kwargs: Any) -> Generator[T, None, None]:
+                for item in generator(*args, **kwargs):
+                    if limiter.total_bytes < limiter.max_dataset_size:
+                        yield item
+                    else:
+                        break
+
+            return wrapped
+
+        self.exit_stack.enter_context(patch.object(ParquetWriter, "_WRITER_CLASS", _TrackedParquetWriter))
+        if isinstance(self.builder, datasets.builder.GeneratorBasedBuilder):
+            self.exit_stack.enter_context(
+                patch.object(self.builder, "_generate_examples", limited_generator(self.builder._generate_examples))
+            )
+        else:
+            self.exit_stack.enter_context(
+                patch.object(self.builder, "_generate_tables", limited_generator(self.builder._generate_tables))
+            )
+        return self
+
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_value: Optional[BaseException],
+        traceback: Optional[TracebackType],
+    ) -> None:
+        return self.exit_stack.close()
+
+
+def list_generated_parquet_files(builder: DatasetBuilder, partial: bool = False) -> List[ParquetFile]:
+    """List the parquet files generated by `builder.download_and_prepare` in the `builder.cache_dir`."""
+    if not builder.info.splits:
+        raise EmptyDatasetError("No split found after generating parquet files")
+    split_dict = builder.info.splits
+    local_parquet_files: List[ParquetFile] = []
+    for split, split_info in split_dict.items():
+        num_shards = len(split_info.shard_lengths) if isinstance(split_info.shard_lengths, list) else 1
+        fname = f"{builder.name}-{split}.parquet"
+        local_parquet_files.extend(
+            [
+                ParquetFile(
+                    local_file=os.path.join(
+                        builder.cache_dir,
+                        fname.replace(
+                            ".parquet",
+                            f"-{shard_idx:05d}-of-{num_shards:05d}.parquet" if num_shards > 1 else ".parquet",
```

Comment:
isn't this implicit dependency on the datasets library implementation?

Reply:
yep, will add a comment